### PR TITLE
Speedy: make thing a bit more private

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 1]
```diff
@@ -199,7 +200,7 @@ object SBuiltin {
   final case object SBDivNumeric extends SBBinaryOpNumeric2(divide)
 
   final case object SBRoundNumeric extends SBuiltin(3) {
-    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+    override def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
```
Comment (rautenrieth-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:203]: Other `SBuiltin` implementations use `final def execute`. Should this one be `final` as well?

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 2]
```diff
@@ -22,11 +22,14 @@ import org.slf4j.LoggerFactory
...
-object Speedy {
+private[lf] object Speedy {
+
+  // fake participant to generate a new transactionSeed when running scenarios
+  private val scenarioServiceParticipant = Ref.ParticipantId.assertFromString("scenario-service")
 
   // Would like these to have zero cost when not enabled. Better still, to be switchable at runtime.
-  val enableInstrumentation: Boolean = false
-  val enableLightweightStepTracing: Boolean = false
+  private val enableInstrumentation: Boolean = false
```
Comment (rautenrieth-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:31]: Can this be even more strict?
```suggestion
  private[this] val enableInstrumentation: Boolean = false
```

Same for other `private` uses below.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 3]
```diff
@@ -241,14 +252,14 @@ object Speedy {
     /** Compute a stack trace from the locations in the continuation stack.
         The last seen location will come last. */
     def stackTrace(): ImmArray[Location] = {
-      val s = new util.ArrayList[Location]
+      val s = ImmArray.newBuilder[Location]
```
Comment (rautenrieth-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:255]: This change does not seem to be related to making things private. I assume `ImmArray` is expected to be more performant than `ArrayList` in this case?

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:255]: `ImmArray` builders are probably not as efficient as `ArrayList`. However in the current case the `ArrayList` is built and then directly converted in `ImmArray`

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala (hunk 2 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:31]: :+1:

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala (hunk 1 above)
Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:203]: Good catch.

**Comment by nickchapman-da** [comment]
I like all the `private` stuff.
But what do we gain from introducing all those `override final` in `SBuiltin` ?
I don't think there is any `override`ing going on...

**Comment by stefanobaghino-da** [comment]
> I like all the `private` stuff.
> But what do we gain from introducing all those `override final` in `SBuiltin` ?
> I don't think there is any `override`ing going on...

I there is not overriding, `scalac` would signal an error where `override` is used improperly.

**Comment by garyverhaegen-da** [comment]
> I there is not overriding, `scalac` would signal an error where `override` is used improperly.

Actually the PR does add `override` markers (used properly, though I thought Scala enforced them 🤷), they're just hidden by default because GitHub is trying to get you to approve changes without reading them.

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -199,7 +200,7 @@ object SBuiltin {
   final case object SBDivNumeric extends SBBinaryOpNumeric2(divide)
 
   final case object SBRoundNumeric extends SBuiltin(3) {
-    def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
+    override override final def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
```
Comment (garyverhaegen-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:203]:
```suggestion
    override final def execute(args: util.ArrayList[SValue], machine: Machine): Unit = {
```

**Comment by stefanobaghino-da** [comment]
> I wish things were private by default in Scala.

`private final` by default. :1st_place_medal:

**Comment by nickchapman-da** [comment]
If there not a useful distinction to be made between (1) overriding (changing) a method definition which already exists in the class being extended, and (2) providing the definition which was abstract in the parent.

I don't see what is gained by marking type-2 cases with a big noisy keyword.

**Comment by remyhaemmerle-da** [comment]
> Let the privatization of DAML LF begin :)

Let the privatization of **Speedy** begin ;)

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala [hunk 5]
```diff
@@ -241,14 +253,14 @@ object Speedy {
...
       kontStack.forEach { k =>
         k match {
-          case KLocation(location) => { s.add(location); () }
+          case KLocation(location) => s += location
           case _ => ()
         }
       }
```
Comment (S11001001) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:262]:
```scala
s ++= kontStack.asScala.iterator.collect { case KLocation(location) => location }
```

If you like. (It's views all the way down.)

Comment (S11001001) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/Speedy.scala:262]: Better yet

```scala
kontStack.asScala.iterator.collect { case KLocation(location) => location }.to[ImmArray]
```
